enterprise_access/apps/api/serializers/content_assignments/assignment.py

```diff
@@ -3,20 +3,76 @@
 """
 import logging
 
+from drf_spectacular.utils import extend_schema_field
 from rest_framework import serializers
 
-from enterprise_access.apps.content_assignments.models import LearnerContentAssignment
+from enterprise_access.apps.content_assignments.constants import (
+    AssignmentActions,
+    AssignmentLearnerStates,
+    AssignmentRecentActionTypes,
+    LearnerContentAssignmentStateChoices
+)
+from enterprise_access.apps.content_assignments.models import LearnerContentAssignment, LearnerContentAssignmentAction
 
 logger = logging.getLogger(__name__)
 
 
+class LearnerContentAssignmentActionSerializer(serializers.ModelSerializer):
+    """
+    A read-only Serializer for responding to requests for ``LearnerContentAssignmentAction`` records.
+    """
+    class Meta:
+        model = LearnerContentAssignmentAction
+        fields = [
+            'uuid',
+            'action_type',
+            'completed_at',
+            'error_reason',
+
+            # Intentionally hide traceback from response, since this is primarily for developers/on-call and could
+            # over-communicate secrets.
+            # 'traceback',
+        ]
+        read_only_fields = fields
+
+
+class LearnerContentAssignmentRecentActionSerializer(serializers.Serializer):
+    """
+    Structured data about the most recent action, meant to power a frontend table column.
+    """
+    action_type = serializers.ChoiceField(
+        help_text='Type of the recent action.',
+        choices=AssignmentRecentActionTypes.CHOICES,
+    )
+    timestamp = serializers.DateTimeField(
+        help_text='Date and time when the action was taken.',
+    )
+
+
 class LearnerContentAssignmentResponseSerializer(serializers.ModelSerializer):
     """
     A read-only Serializer for responding to requests for ``LearnerContentAssignment`` records.
     """
     # This causes the related AssignmentConfiguration to be serialized as a UUID (in the response).
     assignment_configuration = serializers.PrimaryKeyRelatedField(read_only=True)
 
+    actions = LearnerContentAssignmentActionSerializer(
+        help_text='All actions associated with this assignment.',
+        many=True,
+    )
+    recent_action = LearnerContentAssignmentRecentActionSerializer(
+        help_text='Structured data about the most recent action. Meant to power a frontend table column.',
+        source='get_recent_action_data',
```

Comment:
Couldn't these use the annotated fields you added to the queryset?  That would save a couple of queries per row of response data.

Reply:
...or it's possible that an upstream `select_related()` would avoid the additional queries, and that using these functions as the source is fine.